src/pip/_internal/network/auth.py

```diff
@@ -23,11 +25,61 @@
 
 logger = getLogger(__name__)
 
-Credentials = Tuple[str, str, str]
+
+class Credentials(NamedTuple):
+    service_name: str
+    username: str
+    password: str
+
+
+class KeyRingCredential(NamedTuple):
+    username: str
+    password: str
+
+
+class KeyRingCli:
+    """Mirror the parts of keyring's API which pip uses
+
+    Instead of calling the keyring package installed alongside pip
+    we call keyring on the command line which will enable pip to
+    use which ever installation of keyring is available first in
+    PATH.
+    """
+
+    @staticmethod
+    def _quote(string: Optional[str]) -> str:
+        return f"'{string}'"
+
+    def get_credential(
+        self, service_name: str, username: Optional[str]
+    ) -> Optional[KeyRingCredential]:
+        cmd = ["keyring", "get", self._quote(service_name), self._quote(username)]
+        res = subprocess.run(cmd)
+        if res.returncode:
+            return None
+        return KeyRingCredential(username=username, password=res.stdout)
+
+    def set_password(self, service_name: str, username: str, password: str) -> None:
+        cmd = [
+            "echo",
```

Comment:
Windows doesn't have an "echo" command, so this won't work there. Also, the `_quote` function looks like it's doing Unix shell style quoting, which again won't be cross-platform.

You also seem to be assuming `shell=True`, which is very definitely not acceptable - we don't want to risk shell injection attacks here.

Reply:
@pfmoore Yeah, very sorry about that! Was much much too quick a first pass. I've made a few changes which I think make this much more acceptable.